lib/spack/spack/cmd/buildcache.py

```diff
@@ -361,6 +356,8 @@ def install_fn(args):
 
 def list_fn(args):
     """list binary packages available from mirrors"""
+    specs = []
+
```

Comment:
I would say we should die there, but my re-reading of the `BinaryCacheIndex.update()` method today leads me to think if we have no cached indices locally and no mirrors configured (which seems like a valid state to me) this error will be raised when it should not be.  I can dig into that as a part of this PR and verify/fix if you think it makes sense.

The reason I initialized the list to empty is that I ran into some failing unit-test that was hard to diagnose because of the uninitialized variable.

Reply:
Since it's not an error for a mirror to be missing an index, I've added a check when we fetch and cache a binary mirror index: if the index doesn't exist (via `spack.util.web.url_exists()`), then we exit that method early and avoid raising `FetchCacheError`.  Let me know what you think.

Reply:
This was addressed by @scottwittenburg.  `FetchCacheError` probably needs a refactor to be a subclass of `SpackError` but that is for another PR.